draft-yasskin-http-origin-signed-responses.md

```diff
@@ -321,17 +348,485 @@ frames, or could
+`Signed-Response-Headers` headers ({{signed-r-headers}}), and with signatures
+included in one or more `Signature` headers ({{signature-header}}).
+
+The client reconstructs the "significant" parts of the exchange
+({{significant-parts}}) either by taking the sent `application/http2`
+({{application-http2}}) content directly or by building an `application/http2`
+item from the significant request and response headers and the response payload.
+Then it categorizes each `Signature` header as "valid" or "invalid" by
+validating that header's certificate, metadata, and signature against the
+significant parts of the pair ({{signature-validity}}). This validity then
+informs higher-level protocols.
+
+The `Signature` header includes information to let a client fetch assurance a
+signed exchange is still valid, in the face of revoked certificates and
+newly-discovered vulnerabilities. This assurance can be bundled back into the
+signed exchange and forwarded to another client, which won't have to re-fetch
+this validity information for some period of time.
+
+## The Signed-Request-Headers and Signed-Response-Headers Headers ## {#signed-r-headers}
+
+Each of the `Signed-Request-Headers` and `Signed-Response-Headers` headers
+identifies an ordered list of request and response, respectively, header field
+names to include in a signature. These lists of header field names need to be
+consistent across all signatures of an exchange, to avoid forcing higher-level
+protocols to merge the header field lists of valid signatures, so they're
+provided by headers instead of being included in the `Signature` header field.
+
+`Signed-Request-Headers` and `Signed-Response-Headers` are Structured Headers as
+defined by {{!I-D.nottingham-structured-headers}}. Their values MUST be lists
+({{!I-D.nottingham-structured-headers}}, section 4.7) of lowercase strings
+({{!I-D.nottingham-structured-headers}}, section 4.2) naming HTTP headers,
+following the ordering and lowercasing rules from {{!RFC7540}}, section 8.1.2.
+
+## The Signature Header ## {#signature-header}
+
+The `Signature` header conveys a signature for an exchange and information about
+how to determine the authority of and refresh that signature.
+
+The `Signature` header is a Structured Header as defined by
+{{!I-D.nottingham-structured-headers}}. Its value MUST be a dictionary
+({{!I-D.nottingham-structured-headers}}, section 4.7).
+
+The dictionary MUST contain members named "validityUrl", "date", "expires", and
+"sig", and either "certUrl" and "certSha256" members or an "ed25519Key" member.
+The present members MUST have the following values:
+
+"certUrl"
+
+: A string ({{!I-D.nottingham-structured-headers}}, section 4.2) containing a
+  [valid URL string](https://url.spec.whatwg.org/#valid-url-string).
+
+"certSha256"
+
+: Binary content ({{!I-D.nottingham-structured-headers}}, section 4.5) holding
+  the SHA-256 hash of the first certificate found at "certUrl".
+
+"ed25519Key"
```

Comment:
@otherdaniel/@mikewest Does this look like it'll handle what https://github.com/mikewest/signature-based-sri needs? (The algorithm that uses it is at https://wicg.github.io/webpackage/design-signatures/draft-yasskin-http-origin-signed-responses.html#signature-validity)

Reply:
In broad strokes, I think this would support the SRI proposal. I'm not sure about some details, but if we were designing something _just_ for the SRI use case, I think we'd end up with something sufficiently similar to this that we should just squint at it until it's the same. I'm willing to live with some "cruft" (from my feature's perspective) in order to not invent a custom new thing just for my use case. :)

That said, https://github.com/mikewest/signature-based-sri/issues/5#issuecomment-345212824 suggested looking into https://tools.ietf.org/html/draft-cavage-http-signatures-08, which does similar things to this proposal, with a different spelling, and somewhat different characteristics (no certs, for instance). It does tack headers into the message to be signed, which is appealing (binding the resource to a `Content-Type`, for instance, might be valuable).

I know you've thought about that use case. What do you think about supporting it in this way?

Reply:
This is definitely still open for tweaking. For example, I could replace the `ed25519Key=*base64` field with a CSP-style `key=ed25519-base64` [label](https://mnot.github.io/I-D/structured-headers/#label). I didn't because that takes an extra level of parsing to actually check the signature in this spec, but it's certainly possible.

It's plausible that we should have totally separate headers for the "give extra assurance when the physical and logical request are the same" case from the "assert a request/response pair that my normal certificate doesn't cover" case. Even for the extra-assurance case, draft-cavage-http-signatures has some problems:
1. Specifying the algorithm next to the key is bad practice nowadays because it leads to things like https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/.
1. Using a keyId to have a higher-level protocol look up the key instead of having the header specify it leaves room for those protocols to shoot themselves in the foot. I assume SRI would make something like "ed25519-base64" the keyId, so you won't be affected, but it's a problem for adopting the header more broadly.
1. Using the `Digest` header as the only way to sign the body seems risky: we'd have to make sure that every path through the browser notices the presence of a Digest header and avoids using the body until it's checked.
1. I'm still pretty sure leaving the hostname out of the signed data is a mistake, even though it requires extra signatures for testing vs prod. This draft doesn't require people to specify the hostname either, but it doesn't hardcode omitting it.
1. draft-cavage-http-signatures doesn't expire signatures, meaning a single signed resource with a vulnerability forces the server to rotate the key.
1. New things should try to use mnot's structured headers, although that's probably easy to fix.